### ISSUE TITLE
[Performance]: Logprob divergence between vLLM and transformers on a fine-tuned Qwen3.5-VL mobile-use model

### ISSUE THREAD

**Issue by WRWA** [vllm-project/vllm#47425]
### Proposal to improve performance

# Proposal to improve performance

I am not proposing a performance optimization or a new performance-improvement design in this issue.

This report is about an observed accuracy / logprob discrepancy between vLLM and HuggingFace transformers when serving a fine-tuned Qwen3.5-VL mobile-use model. The goal is to share the observation and ask whether this type of divergence is known or expected for Qwen3.5-VL in vLLM.

N/A.

### Report of performance regression

# Report of performance regression

## Summary

I observed a consistent performance gap when evaluating a fine-tuned Qwen3.5-VL model with HuggingFace transformers and vLLM.

The model was fine-tuned on my own private data, so exact reproduction may be difficult. However, the usage pattern is based on Qwen’s public `mobile_use` agent example: the model emits tool calls such as `click`, `swipe`, and `terminate` in the following style:

```text
<tool_call>
{"name":"mobile_use","arguments":{"action":"
```

In several cases, transformers and vLLM produced almost identical logprobs for the initial structured tool-call prefix, but diverged sharply at the first action token after `"action":"`. The difference was sometimes large enough to flip the greedy argmax and change the emitted action.

I understand that vLLM does not guarantee bitwise parity or exact reproducibility with transformers. I am sharing this because the divergence appears concentrated at a task-critical action-token decision point rather than being uniformly distributed across the generated sequence.

## Additional comparison with Gemma

For comparison, I ran similar evaluations with a fine-tuned Gemma model.

With Gemma, the results across transformers and the serving engine were within a range that I would consider acceptable given normal environment and engine differences. This held for both bf16 and 4-bit quantized variants. The observed difference was less than about 0.8 percentage points.

In contrast, with Qwen3.5-VL, the performance gap appeared specifically when running the model with vLLM. The issue was present regardless of whether the model was bf16 or 4-bit quantized. The observed gap was roughly 3 percentage points, which is substantially larger than what I observed with Gemma.

This makes the issue look specific to the Qwen + vLLM combination rather than a general evaluation noise or serving-engine variance issue.

## Experiment

I compared transformers and vLLM on the same evaluation inputs.

For transformers, I ran the model directly in Python and recorded the top candidate tokens and logprobs at each generation step.

For vLLM, I queried a running vLLM server with `logprobs=True` and `top_logprobs=20`, then compared the returned top candidates and logprobs.

To reduce preprocessing/tokenization confounds, I focused on cases where the transformers prompt length matched vLLM’s `usage.prompt_tokens`. In those matched-token-count cases, both engines generated the same structured prefix up to the action field, then diverged at the action-token decision point.

## Observation 1: the structured prefix is almost identical

For matched prompt-token-count cases, both engines generated the same prefix tokens up to the action value.

The common prefix was:

```text
<tool_call>
{"name":"mobile_use","arguments":{"action":"
```

Across the identical-prefix region, the per-token absolute logprob difference was very small:

```text
median |Δ logprob| = 0
mean   |Δ logprob| ≈ 0.007
max    |Δ logprob| < 0.6 nat
```

This suggests that the discrepancy is not spread uniformly across the generated sequence. It is concentrated at the action-token decision point.

## Observation 2: the action-token distribution can invert

One representative case:

```text
prompt tokens: transformers plen = vLLM usage.prompt_tokens = 1521
divergence step: first action token after "action":"
```

At the decision token:

| token | transformers logprob | vLLM logprob | Δ vLLM - transformers |
|---|---:|---:|---:|
| `click` | ~0.000 | -0.916 | -0.916 |
| `terminate` | -8.375 | -0.541 | +7.834 |
| `sw` | -11.500 | -4.041 | +7.459 |

In this case, transformers is almost certain about `click`, while vLLM assigns higher probability to `terminate`. With greedy decoding, this changes the emitted action.

Another representative case showed the same type of inversion:

| token | transformers logprob | vLLM logprob | Δ vLLM - transformers |
|---|---:|---:|---:|
| `click` | -0.017 | -3.898 | -3.881 |
| `terminate` | -4.267 | -0.023 | +4.244 |

Here, both engines are highly confident, but in opposite directions.

## Ablation

| variable changed | result |
|---|---|
| model family comparison | Gemma showed less than ~0.8 percentage points difference across engines / precision settings, while Qwen3.5-VL showed roughly ~3 percentage points difference when served with vLLM |
| vLLM attention backend: `FLASH_ATTN` vs `FLASHINFER` | no material difference in aggregate accuracy |
| quantization | the detailed logprob table above is from GPTQ, but the same lower-performance pattern under vLLM was also observed before quantization / in bf16 |
| tensor parallelism | not used; single GPU, TP=1 |

## Note on vLLM benchmark scripts

I understand that vLLM provides benchmark scripts under `benchmarks/`. In this case, the observed regression is not throughput or latency-related. It is an accuracy / logprob divergence in a multimodal tool-use evaluation setup, so the standard performance benchmark scripts may not directly capture the issue.

## Question

Is this kind of multi-nat action-token logprob divergence a known or expected behavior for Qwen3.5-VL in vLLM, especially around the `mobile_use`-style action decision token?

I can provide the logprob comparison script and numeric evidence if useful, but I cannot share the fine-tuned weights or private evaluation images.

### Misc discussion on performance

# Misc discussion on performance

The main discussion point is that this looks like an accuracy / logprob behavior difference rather than a throughput or latency issue.

In my evaluation, the issue appeared specifically with the Qwen3.5-VL + vLLM combination. A fine-tuned Gemma model evaluated under similar conditions showed less than about 0.8 percentage points difference across engines / precision settings, which I would consider acceptable given normal environment and serving differences.

For Qwen3.5-VL, the observed gap was roughly 3 percentage points when served with vLLM. This happened for both bf16 and 4-bit quantized variants, although the detailed step-wise logprob examples I have are from the GPTQ 4-bit model.

The most notable pattern is that the logprob difference is not uniformly distributed across the generated sequence. The initial structured tool-call prefix is almost identical between transformers and vLLM, but the first action token after `"action":"` can show a multi-nat difference large enough to flip greedy decoding.

I am mainly trying to understand whether this is a known or expected behavior for Qwen3.5-VL in vLLM, particularly in `mobile_use`-style action generation.

### Your current environment (if you think it is necessary)

| item | value |
|---|---|
| vLLM | `0.23.0` |
| transformers | `5.12.1` |
| torch | `2.11.0+cu130` |
| CUDA | `13.0` |
| GPU | RTX 4070 12 GB |
| model architecture | Qwen3.5-VL |
| quantization for the logprob experiment | GPTQ 4-bit, `gptqmodel 7.1.0` |
| decoding | greedy, `temperature=0`, `top_p=1`; transformers `do_sample=False` |
| vLLM server flags | `--max-model-len 12288 --gpu-memory-utilization 0.45` |
| tensor parallelism | single GPU, TP=1 |

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by aoshen02** [comment]
I ran a controlled HF-vs-vLLM fixed-token logprob check for this report.

**Images tested**
- `vllm/vllm-openai:v0.23.0` (digest `sha256:6d8429e38e3747723ca07ee1b17972e09bb9c51c4032b266f24fb1cc3b22ed8f`), which contains vLLM 0.23.0, transformers 5.12.0, and torch 2.11.0+cu130.
- `vllm/vllm-openai:nightly` (digest `sha256:212a1bd7b4267c604408d17dc0048ef152101bc67fbe6ba8567899fc1f227bcd`), vLLM `0.23.1rc1.dev1327+gf25953cc5`.

I used public `Qwen/Qwen3.5-0.8B`, BF16, TP=1, eager mode, greedy decoding, and scored the first action candidates `click`, `terminate`, and `sw` by forcing the token as the final prompt token (rather than comparing scores after generation has already diverged).

| Control | Prompt | Result |
|---|---:|---|
| nightly | text, 1521 tokens | same greedy token; max abs Δlogprob = 0.05496 nat |
| nightly | one real image, 32 text tokens | same greedy `click`; max = 0.30258 nat |
| nightly | one real image, 1521 text tokens | same greedy `click`; max = 0.17232 nat |
| **v0.23.0 + transformers 5.12** | **one real image, 1521 text tokens** | **same greedy `click`; max = 0.19744 nat** |

For both image controls, I rendered the Qwen chat template with HF's `AutoProcessor`, passed the same image to vLLM, and verified vLLM's complete processed `prompt_token_ids` exactly matched HF's input IDs. Thus the public control exercises the vision encoder and the 1521-token GDN/hybrid prefill path, not only text tokenization.

This does **not** reproduce the 4–8 nat inversion, including with the issue's vLLM/transformers version family. It makes a general v0.23.0/nightly sampler, returned-logprob, or public Qwen3.5-VL image-path discrepancy unlikely. The remaining differentiators are specific to the unavailable private checkpoint/request: exact fine-tuned weights/revision and conversion, image preprocessing outputs (`pixel_values`, grid metadata), M-RoPE position IDs, and the exact chat-template/token-ID sequence. Matching only `usage.prompt_tokens` proves a count, not equality of those multimodal inputs.

If sharing weights/images is impossible, a single failing case could still be diagnosed with an export of: input IDs, selected-token logits/logprobs, `pixel_values` checksum/shape/dtype, image/video grid metadata, M-RoPE position IDs, chat template, and checkpoint revision/conversion details.
